File: 11-sistema-notificaciones-go/internal/dominio/entidad/usuario.go
package entidad

import (
	"time"
	"gorm.io/gorm"
)

// EstadoUsuario define los estados de un usuario
type EstadoUsuario string

const (
	EstadoActivo       EstadoUsuario = "activo"
	EstadoInactivo     EstadoUsuario = "inactivo"
	EstadoSuspendido   EstadoUsuario = "suspendido"
	EstadoPendiente    EstadoUsuario = "pendiente"
)

// RolUsuario define los roles de un usuario
type RolUsuario string

const (
	RolAdministrador  RolUsuario = "administrador"
	RolModerador      RolUsuario = "moderador"
	RolUsuario        RolUsuario = "usuario"
	RolInvitado       RolUsuario = "invitado"
)

// Usuario representa un usuario en el sistema
type Usuario struct {
	ID                uint           `json:"id" gorm:"primaryKey"`
	NombreUsuario     string         `json:"nombre_usuario" gorm:"uniqueIndex;not null;size:50"`
	CorreoElectronico string         `json:"correo_electronico" gorm:"uniqueIndex;not null;size:255"`
	Nombre            string         `json:"nombre" gorm:"not null;size:100"`
	Apellido          string         `json:"apellido" gorm:"not null;size:100"`
	Telefono          string         `json:"telefono" gorm:"size:20"`
	Estado            EstadoUsuario  `json:"estado" gorm:"not null;size:50;default:'activo'"`
	Rol               RolUsuario     `json:"rol" gorm:"not null;size:50;default:'usuario'"`
	CorreoVerificado  bool           `json:"correo_verificado" gorm:"default:false"`
	TelefonoVerificado bool          `json:"telefono_verificado" gorm:"default:false"`
	UltimoAcceso      *time.Time     `json:"ultimo_acceso"`
	FechaCreacion     time.Time      `json:"fecha_creacion" gorm:"autoCreateTime"`
	FechaActualizacion time.Time     `json:"fecha_actualizacion" gorm:"autoUpdateTime"`
	FechaEliminacion  gorm.DeletedAt `json:"fecha_eliminacion" gorm:"index"`
	
	// Relaciones
	Notificaciones    []Notificacion `json:"notificaciones" gorm:"foreignKey:UsuarioID"`
	Canales           []Canal        `json:"canales" gorm:"many2many:usuario_canales;"`
}

// NuevoUsuario crea una nueva instancia de Usuario
func NuevoUsuario(nombreUsuario, correoElectronico, nombre, apellido string) *Usuario {
	return &Usuario{
		NombreUsuario:     nombreUsuario,
		CorreoElectronico: correoElectronico,
		Nombre:            nombre,
		Apellido:          apellido,
		Estado:            EstadoActivo,
		Rol:               RolUsuario,
		CorreoVerificado:  false,
		TelefonoVerificado: false,
	}
}

// ObtenerNombreCompleto retorna el nombre completo del usuario
func (u *Usuario) ObtenerNombreCompleto() string {
	return u.Nombre + " " + u.Apellido
}

// EstaActivo verifica si el usuario está activo
func (u *Usuario) EstaActivo() bool {
	return u.Estado == EstadoActivo
}

// EsAdministrador verifica si el usuario es administrador
func (u *Usuario) EsAdministrador() bool {
	return u.Rol == RolAdministrador
}

// EsModerador verifica si el usuario es moderador
func (u *Usuario) EsModerador() bool {
	return u.Rol == RolModerador
}

// ActualizarUltimoAcceso actualiza la fecha del último acceso
func (u *Usuario) ActualizarUltimoAcceso() {
	ahora := time.Now()
	u.UltimoAcceso = &ahora
}

// VerificarCorreo marca el correo como verificado
func (u *Usuario) VerificarCorreo() {
	u.CorreoVerificado = true
}

// VerificarTelefono marca el teléfono como verificado
func (u *Usuario) VerificarTelefono() {
	u.TelefonoVerificado = true
}

// Activar activa el usuario
func (u *Usuario) Activar() {
	u.Estado = EstadoActivo
}

// Desactivar desactiva el usuario
func (u *Usuario) Desactivar() {
	u.Estado = EstadoInactivo
}

// Suspender suspende el usuario
func (u *Usuario) Suspender() {
	u.Estado = EstadoSuspendido
}

// CambiarRol cambia el rol del usuario
func (u *Usuario) CambiarRol(nuevoRol RolUsuario) {
	u.Rol = nuevoRol
}

// Validar valida el usuario
func (u *Usuario) Validar() error {
	if u.NombreUsuario == "" {
		return NewErrorValidacion("Nombre de usuario es requerido")
	}
	if u.CorreoElectronico == "" {
		return NewErrorValidacion("Correo electrónico es requerido")
	}
	if u.Nombre == "" {
		return NewErrorValidacion("Nombre es requerido")
	}
	if u.Apellido == "" {
		return NewErrorValidacion("Apellido es requerido")
	}
	return nil
}

File: 06-arquitectura-microservicios-java/microservicios/servicio-usuarios/src/main/java/com/portfolio/arquitectura/usuarios/dominio/entidad/Usuario.java
package com.portfolio.arquitectura.usuarios.dominio.entidad;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.persistence.*;
import java.time.LocalDateTime;
import java.util.UUID;

/**
 * Entidad Usuario - Capa de Dominio
 * 
 * Representa un usuario en el sistema de e-commerce.
 * Implementa los principios de DDD con agregados bien definidos.
 * 
 * @author Gabriel - Arquitecto de Software
 */
@Entity
@Table(name = "usuarios")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Usuario {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private UUID id;

    @Column(unique = true, nullable = false)
    private String correoElectronico;

    @Column(unique = true, nullable = false)
    private String nombreUsuario;

    @Column(nullable = false)
    private String contrasena;

    @Column(name = "nombre", nullable = false)
    private String nombre;

    @Column(name = "apellido", nullable = false)
    private String apellido;

    @Column(name = "numero_telefono")
    private String numeroTelefono;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private EstadoUsuario estado;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private RolUsuario rol;

    @Column(name = "correo_verificado")
    private Boolean correoVerificado;

    @Column(name = "fecha_creacion", nullable = false)
    private LocalDateTime fechaCreacion;

    @Column(name = "fecha_actualizacion")
    private LocalDateTime fechaActualizacion;

    @Column(name = "ultimo_acceso")
    private LocalDateTime ultimoAcceso;

    @PrePersist
    protected void alCrear() {
        fechaCreacion = LocalDateTime.now();
        fechaActualizacion = LocalDateTime.now();
        if (estado == null) {
            estado = EstadoUsuario.ACTIVO;
        }
        if (rol == null) {
            rol = RolUsuario.CLIENTE;
        }
        if (correoVerificado == null) {
            correoVerificado = false;
        }
    }

    @PreUpdate
    protected void alActualizar() {
        fechaActualizacion = LocalDateTime.now();
    }

    /**
     * Verifica si el usuario está activo
     */
    public boolean estaActivo() {
        return EstadoUsuario.ACTIVO.equals(this.estado);
    }

    /**
     * Verifica si el usuario es administrador
     */
    public boolean esAdministrador() {
        return RolUsuario.ADMINISTRADOR.equals(this.rol);
    }

    /**
     * Verifica si el usuario es vendedor
     */
    public boolean esVendedor() {
        return RolUsuario.VENDEDOR.equals(this.rol);
    }

    /**
     * Actualiza el último acceso
     */
    public void actualizarUltimoAcceso() {
        this.ultimoAcceso = LocalDateTime.now();
    }

    /**
     * Marca el correo como verificado
     */
    public void verificarCorreo() {
        this.correoVerificado = true;
    }

    /**
     * Desactiva el usuario
     */
    public void desactivar() {
        this.estado = EstadoUsuario.INACTIVO;
    }

    /**
     * Activa el usuario
     */
    public void activar() {
        this.estado = EstadoUsuario.ACTIVO;
    }
}
